File: ios-native/Leyne/AnalyticsService.swift
//  AnalyticsService.swift
//
//  Thin, typed wrapper over Firebase Analytics — the app's high-signal product
//  events. Centralising the Firebase import here means the rest of the app logs
//  via a typed API (`AnalyticsService.log(.stopViewed(...))`) and never imports
//  FirebaseAnalytics directly, so event names + parameters have one source of
//  truth that the Android side mirrors. `app_open` is logged automatically by the
//  SDK, so it is deliberately not represented here.
//
//  NOTE: ad revenue (`ad_impression`) is NOT logged here — the Google Mobile Ads
//  SDK auto-logs it once the AdMob↔Firebase link is active. Don't re-add a manual
//  paidEventHandler logger, or impressions double-count in GA4.
//
//  Firebase is configured at launch in `LeyneApp` (LeyneAppDelegate). When the
//  GoogleService-Info.plist is absent (e.g. a fork / CI without the config),
//  configure() is skipped and these calls degrade to no-ops — never a crash.

import Foundation
import FirebaseAnalytics

enum AnalyticsService {

    // MARK: - Product events

    /// The seven high-signal events from the Phase 0 plan (minus auto `app_open`).
    enum Event {
        case stopViewed(code: String, kind: StopKind)
        case alertSet(kind: String, busNo: String)
        case favouriteAdded(kind: FavKind)
        case searchPerformed
        case onboardingCompleted
        case notificationTapped(kind: String)

        enum StopKind: String { case bus, mrt }
        enum FavKind: String { case stop, service }

        var name: String {
            switch self {
            case .stopViewed:          return "stop_viewed"
            case .alertSet:            return "alert_set"
            case .favouriteAdded:      return "favourite_added"
            case .searchPerformed:     return "search_performed"
            case .onboardingCompleted: return "onboarding_completed"
            case .notificationTapped:  return "notification_tapped"
            }
        }

        var parameters: [String: Any]? {
            switch self {
            case let .stopViewed(code, kind):
                return ["stop_code": code, "kind": kind.rawValue]
            case let .alertSet(kind, busNo):
                return ["kind": kind, "bus_no": busNo]
            case let .favouriteAdded(kind):
                return ["kind": kind.rawValue]
            case .searchPerformed, .onboardingCompleted:
                return nil
            case let .notificationTapped(kind):
                return ["kind": kind.isEmpty ? "unknown" : kind]
            }
        }
    }

    /// Log a product event. Safe to call before Firebase is configured (no-op).
    static func log(_ event: Event) {
        Analytics.logEvent(event.name, parameters: event.parameters)
    }
}

File: lib/services/analytics_service.dart
// AnalyticsService — thin, typed wrapper over Firebase Analytics for the app's
// high-signal product events. Centralising the FirebaseAnalytics import here
// means the rest of the app logs via a typed API and never touches the SDK
// directly, so event names + parameters have one source of truth. This mirrors
// the iOS AnalyticsService.swift one-to-one (same event names, same parameter
// keys) so GA4 reports both platforms together.
//
// `app_open` is logged automatically by the SDK, so it is deliberately not
// represented here.
//
// NOTE: ad revenue (`ad_impression`) is NOT logged here — the Google Mobile Ads
// SDK auto-logs it once the AdMob↔Firebase link is active. Do NOT add a manual
// OnPaidEvent / paidEventHandler logger, or impressions double-count in GA4.
// (Matches the iOS decision — see AnalyticsService.swift.)
//
// Firebase is configured at launch in main.dart, guarded on the presence of
// google-services.json (the file is git-ignored — forks / CI / pre-setup builds
// skip it). When Firebase isn't configured, [_ready] stays false and every call
// here degrades to a no-op — never a crash.

import 'package:firebase_analytics/firebase_analytics.dart';
import 'package:flutter/foundation.dart';

enum StopKind { bus, mrt }

enum FavKind { stop, service }

class AnalyticsService {
  AnalyticsService._();

  /// Flipped true by [markReady] once Firebase.initializeApp() has succeeded in
  /// main.dart. Until then (and forever, on a build without google-services.json)
  /// every log call no-ops. Mirrors iOS configure()-skipped behaviour.
  static bool _ready = false;

  static FirebaseAnalytics? _analytics;

  /// Called from main.dart after a successful Firebase.initializeApp(). Safe to
  /// call more than once.
  static void markReady() {
    if (_ready) return;
    _ready = true;
    _analytics = FirebaseAnalytics.instance;
  }

  // ─── Product events (mirror iOS AnalyticsService.Event) ───────────────────

  /// A stop / station detail was opened. [code] is the bus stop code or the
  /// station's first code; [kind] distinguishes bus vs MRT.
  static void stopViewed({required String code, required StopKind kind}) =>
      _log('stop_viewed', {'stop_code': code, 'kind': kind.name});

  /// A notification alert was set for a service.
  static void alertSet({required String kind, required String busNo}) =>
      _log('alert_set', {'kind': kind, 'bus_no': busNo});

  /// A stop or service was added to favourites/saved.
  static void favouriteAdded(FavKind kind) =>
      _log('favourite_added', {'kind': kind.name});

  /// A search was performed.
  static void searchPerformed() => _log('search_performed', null);

  /// Onboarding was completed (or skipped — same routing effect).
  static void onboardingCompleted() => _log('onboarding_completed', null);

  /// A notification was tapped (a strong value signal for retention analysis).
  static void notificationTapped(String kind) =>
      _log('notification_tapped', {'kind': kind.isEmpty ? 'unknown' : kind});

  // ─── Internal ─────────────────────────────────────────────────────────────

  /// Fire-and-forget. No-op before Firebase is ready; never throws into the
  /// caller (analytics must never break a user flow).
  static void _log(String name, Map<String, Object>? parameters) {
    final a = _analytics;
    if (!_ready || a == null) return;
    a.logEvent(name: name, parameters: parameters).catchError((Object e) {
      if (kDebugMode) {
        // ignore: avoid_print
        print('[analytics] logEvent($name) failed: $e');
      }
    });
  }
}
